File: src/utils_cpp/shapes/Box.cpp
#include "Box.hpp"

Box::Box() {}
Box::Box(vec center_, vec borders_) : Shape(center_), borders(borders_) {}

Box::Box(const Box& lhs) : Shape(lhs), borders(lhs.borders) {}

Box& Box::operator=(const Box& lhs) {
    Box t(lhs);
    std::swap(center, t.center);
    std::swap(borders, t.borders);

    return *this;
}

Box::Box(Box&& rhs) : Shape(rhs), borders(rhs.borders)  {
    rhs.center = vec();
    rhs.borders = vec();
}

Box& Box::operator=(Box&& rhs) {
    Box t(std::move(rhs));
    std::swap(center, t.center);
    std::swap(borders, t.borders);

    return *this;
}


double Box::get_volume() const {
    return borders[0] * borders[1] * borders[2];
}

double Box::get_surface() const {
    return 2 * (borders[0] * borders[1] + borders[1] * borders[2] + borders[0] * borders[2]);
}

bool Box::check_point(vec point) const {
    vec shifted_pos = point - center;
    return (abs(shifted_pos[0]) < borders[0] / 2) &&
           (abs(shifted_pos[1]) < borders[1] / 2) &&
           (abs(shifted_pos[2]) < borders[2] / 2);
}

vec Box::generate_point(std::mt19937& gen) const {
    std::uniform_real_distribution<double> dist_x(0, borders[0]);
    std::uniform_real_distribution<double> dist_y(0, borders[1]);
    std::uniform_real_distribution<double> dist_z(0, borders[2]);

    return vec({dist_x(gen), dist_y(gen), dist_z(gen)});
}

File: panda/geom/Box.py
from typing import Union

import numpy as np

from ..utils import validate_list_and_array
from .Shape import Shape


class Box(Shape):
    def __init__(self, center: Union[list, np.array], borders: Union[list, np.array]):
        super().__init__()

        self.center = validate_list_and_array(center)
        self.borders = validate_list_and_array(borders)

    def get_volume(self) -> float:
        return np.prod(self.borders)

    def get_surface(self) -> float:
        return 2 * (
            self.borders[0] * self.borders[1]
            + self.borders[1] * self.borders[2]
            + self.borders[0] * self.borders[2]
        )

    def check_point(self, point) -> bool:
        return np.prod(np.abs(point - self.center) < self.borders / 2)

    def generate_point(self) -> np.array:
        return self.borders * np.random.uniform(-0.5, 0.5, 3) + self.center
